File: backend/app/api/launcher.py
import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
import subprocess
import shutil
import platform

from app.api.auth import get_current_user_id

logger = logging.getLogger(__name__)

api_router = APIRouter(prefix="/launcher", tags=["launcher"])

# 注意（M2 多租户复核）：launcher 在「后端机器」上以子进程方式启动 Cursor/Codex，
# 属本地能力（迁移后归桌面壳/本地代理，M5）。在迁移完成前，云端形态下必须要求
# 登录鉴权，杜绝未授权用户触发服务器进程执行。前端经 apiClient 自动附带 Bearer。
# 进一步建议：cloud 模式应整体禁用本路由（见 M2 文档「残留风险」）。

SUPPORTED_TOOLS = (
    "cursor",
    "codex-cli",
    "codex-app",
    "windsurf",
    "claude-code",
    "claude-app",
)
IS_WINDOWS = platform.system() == "Windows"

TOOL_LABELS = {
    "cursor": "Cursor",
    "codex-cli": "Codex CLI",
    "codex-app": "Codex App",
    "windsurf": "Windsurf",
    "claude-code": "Claude Code",
    "claude-app": "Claude",
}

# 交互式 CLI 工具（新终端窗口启动）；其余按 GUI 应用后台启动。
TERMINAL_TOOLS = ("codex-cli", "claude-code")


class LaunchRequest(BaseModel):
    tool: str
    project_path: str = ""


class ToolInfo(BaseModel):
    id: str
    label: str
    available: bool
    mode: str
    description: str


def _find_executable(*candidates: str) -> Optional[str]:
    """Return the first candidate found in PATH, or None."""
    for name in candidates:
        path = shutil.which(name)
        if path:
            return path
    return None


_appx_cache: dict[str, Optional[str]] = {}


def _find_appx_app(package_pattern: str) -> Optional[str]:
    """在 Windows 上查找 MSIX/AppX 安装的应用，返回 shell:AppsFolder URI。"""
    if not IS_WINDOWS:
        return None
    if package_pattern in _appx_cache:
        return _appx_cache[package_pattern]
    try:
        result = subprocess.run(
            ["powershell", "-NoProfile", "-Command",
             f"Get-StartApps | Where-Object {{ $_.Name -like '{package_pattern}' }} "
             "| Select-Object -First 1 -ExpandProperty AppID"],
            capture_output=True, text=True, timeout=15,
        )
        app_id = result.stdout.strip()
        logger.info(f"[launcher] AppX 查询结果: rc={result.returncode}, id={app_id!r}")
        if app_id and result.returncode == 0:
            uri = f"shell:AppsFolder\\{app_id}"
            _appx_cache[package_pattern] = uri
            return uri
    except subprocess.TimeoutExpired:
        logger.warning(f"[launcher] 查找 AppX 应用超时: {package_pattern}")
    except Exception as e:
        logger.warning(f"[launcher] 查找 AppX 应用失败: {e}")
    _appx_cache[package_pattern] = None
    return None


def _resolve_command(tool: str) -> list[str]:
    """Resolve the CLI command for the given tool on the current platform."""

    if tool == "cursor":
        candidates = ("cursor.cmd", "cursor") if IS_WINDOWS else ("cursor",)
        exe = _find_executable(*candidates)
        if exe:
            return [exe]
        raise FileNotFoundError(
            "cursor 命令未找到，请确认 Cursor 已安装且在 PATH 中"
        )

    if tool == "codex-cli":
        exe = _find_executable("codex.cmd", "codex") if IS_WINDOWS else _find_executable("codex")
        if exe:
            return [exe]
        raise FileNotFoundError(
            "codex 命令未找到，请确认 Codex CLI 已安装 (npm i -g @openai/codex)"
        )

    if tool == "codex-app":
        if IS_WINDOWS:
            exe = _find_executable("codex-app.cmd", "codex-app", "Codex.exe")
            if exe:
                return [exe]
            appx_uri = _find_appx_app("Codex")
            if appx_uri:
                return ["explorer.exe", appx_uri]
        else:
            exe = _find_executable("codex-app", "Codex")
            if exe:
                return [exe]
        raise FileNotFoundError(
            "Codex App 未找到，请确认 Codex 桌面应用已安装"
        )

    if tool == "windsurf":
        if IS_WINDOWS:
            exe = _find_executable("windsurf.cmd", "windsurf", "Windsurf.exe")
            if exe:
                return [exe]
            appx_uri = _find_appx_app("Windsurf")
            if appx_uri:
                return ["explorer.exe", appx_uri]
        else:
            exe = _find_executable("windsurf", "Windsurf")
            if exe:
                return [exe]
        raise FileNotFoundError(
            "windsurf 命令未找到，请确认 Windsurf 已安装且在 PATH 中"
        )

    if tool == "claude-code":
        exe = _find_executable("claude.cmd", "claude") if IS_WINDOWS else _find_executable("claude")
        if exe:
            return [exe]
        raise FileNotFoundError(
            "claude 命令未找到，请确认 Claude Code 已安装 (npm i -g @anthropic-ai/claude-code)"
        )

    if tool == "claude-app":
        if IS_WINDOWS:
            exe = _find_executable("claude-app.cmd", "claude-app", "Claude.exe")
            if exe:
                return [exe]
            appx_uri = _find_appx_app("Claude")
            if appx_uri:
                return ["explorer.exe", appx_uri]
        else:
            exe = _find_executable("claude-app", "Claude")
            if exe:
                return [exe]
        raise FileNotFoundError(
            "Claude App 未找到，请确认 Claude 桌面应用已安装"
        )

    raise ValueError(f"不支持的工具: {tool}")


def _launch_background(cmd: list[str]) -> None:
    """以后台静默方式启动（用于桌面 GUI 应用）"""
    kwargs: dict = dict(stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    if IS_WINDOWS:
        kwargs["creationflags"] = (
            subprocess.DETACHED_PROCESS | subprocess.CREATE_NO_WINDOW
        )
        kwargs["shell"] = True
    subprocess.Popen(cmd, **kwargs)


def _launch_terminal(cmd: list[str]) -> None:
    """在新终端窗口中启动（用于 CLI 交互式工具）"""
    if IS_WINDOWS:
        subprocess.Popen(
            ["start", "cmd", "/k"] + cmd,
            shell=True,
            creationflags=subprocess.CREATE_NEW_CONSOLE,
        )
    elif platform.system() == "Darwin":
        script = " ".join(cmd)
        subprocess.Popen(
            ["osascript", "-e", f'tell app "Terminal" to do script "{script}"']
        )
    else:
        for term in ("x-terminal-emulator", "gnome-terminal", "xterm"):
            if shutil.which(term):
                subprocess.Popen([term, "-e"] + cmd)
                return
        subprocess.Popen(cmd)


@api_router.get("/tools")
async def list_tools(user_id: str = Depends(get_current_user_id)):
    """返回所有支持的工具及其可用状态"""
    tools = []
    for tool_id in SUPPORTED_TOOLS:
        try:
            _resolve_command(tool_id)
            available = True
        except FileNotFoundError:
            available = False

        if tool_id == "codex-cli":
            mode, desc = "terminal", "在终端中启动 Codex CLI 交互式对话"
        elif tool_id == "codex-app":
            mode, desc = "app", "启动 Codex 桌面应用"
        elif tool_id == "windsurf":
            mode, desc = "app", "启动 Windsurf IDE"
        elif tool_id == "claude-code":
            mode, desc = "terminal", "在终端中启动 Claude Code 交互式对话"
        elif tool_id == "claude-app":
            mode, desc = "app", "启动 Claude 桌面应用"
        else:
            mode, desc = "app", "启动 Cursor IDE"

        tools.append(ToolInfo(
            id=tool_id,
            label=TOOL_LABELS[tool_id],
            available=available,
            mode=mode,
            description=desc,
        ))
    return {"tools": tools}


@api_router.post("/launch")
async def launch_tool(
    data: LaunchRequest, user_id: str = Depends(get_current_user_id)
):
    """启动工具 (Cursor / Codex CLI / Codex App / Windsurf)"""
    if data.tool not in SUPPORTED_TOOLS:
        raise HTTPException(
            status_code=400,
            detail=f"不支持的工具: {data.tool}，可选: {', '.join(SUPPORTED_TOOLS)}",
        )

    try:
        cmd = _resolve_command(data.tool)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc))

    if data.project_path:
        cmd.append(data.project_path)

    label = TOOL_LABELS[data.tool]
    is_terminal = data.tool in TERMINAL_TOOLS
    try:
        if is_terminal:
            _launch_terminal(cmd)
        else:
            _launch_background(cmd)
    except OSError as exc:
        raise HTTPException(
            status_code=500,
            detail=f"启动 {label} 失败: {exc}",
        )

    suffix = f"，项目路径: {data.project_path}" if data.project_path else ""
    return {
        "status": "launched",
        "tool": data.tool,
        "mode": "terminal" if is_terminal else "app",
        "message": f"{label} 已成功启动{suffix}",
    }

File: desktop/src/main/launcher.ts
import { spawn, spawnSync } from "node:child_process";
import type {
  LauncherLaunchRequest,
  LauncherLaunchResponse,
  LauncherToolId,
  LauncherToolInfo,
} from "../shared/types";

/**
 * 一键启动工具（方案 B M5-a / 决策 D）。
 *
 * 在 Electron 主进程用 TS 重做原 `backend/app/api/launcher.py` 的命令解析与启动
 * （cursor / codex-cli / codex-app / windsurf）——cloud 形态已下线后端 /launcher 路由，
 * 桌面壳在本机直接启动。经 IPC 暴露给渲染层（见 ipc.ts）。
 *
 * 命令解析口径与 launcher.py 一致：
 *   · cursor   : which(cursor.cmd/cursor) → 后台静默启动；
 *   · codex-cli: which(codex.cmd/codex)   → 新终端窗口启动（交互式）；
 *   · codex-app: which(codex-app.cmd/codex-app/Codex.exe) 或 AppX(Get-StartApps) → 后台启动；
 *   · windsurf : which(windsurf.cmd/windsurf/Windsurf.exe) 或 AppX(Get-StartApps) → 后台启动。
 */

const IS_WINDOWS = process.platform === "win32";
const IS_MAC = process.platform === "darwin";

const SUPPORTED_TOOLS: LauncherToolId[] = [
  "cursor",
  "codex-cli",
  "codex-app",
  "windsurf",
  "claude-code",
  "claude-app",
];

const TOOL_LABELS: Record<LauncherToolId, string> = {
  cursor: "Cursor",
  "codex-cli": "Codex CLI",
  "codex-app": "Codex App",
  windsurf: "Windsurf",
  "claude-code": "Claude Code",
  "claude-app": "Claude",
};

/** 交互式 CLI 工具（新终端窗口启动）；其余按 GUI 应用后台启动。 */
const TERMINAL_TOOLS: ReadonlySet<LauncherToolId> = new Set([
  "codex-cli",
  "claude-code",
]);

/** 在 PATH 中查找首个可用可执行文件，返回绝对路径或 null（对齐 shutil.which）。 */
function which(...candidates: string[]): string | null {
  for (const name of candidates) {
    try {
      const cmd = IS_WINDOWS ? "where" : "which";
      const r = spawnSync(cmd, [name], { encoding: "utf-8", windowsHide: true });
      if (r.status === 0 && r.stdout) {
        const first = r.stdout
          .split(/\r?\n/)
          .map((s) => s.trim())
          .filter(Boolean)[0];
        if (first) return first;
      }
    } catch {
      /* try next candidate */
    }
  }
  return null;
}

const appxCache = new Map<string, string | null>();

/** Windows：查 MSIX/AppX 安装的应用，返回 shell:AppsFolder URI（对齐 _find_appx_app）。 */
function findAppxApp(pattern: string): string | null {
  if (!IS_WINDOWS) return null;
  if (appxCache.has(pattern)) return appxCache.get(pattern) ?? null;
  try {
    const r = spawnSync(
      "powershell",
      [
        "-NoProfile",
        "-Command",
        `Get-StartApps | Where-Object { $_.Name -like '${pattern}' } | Select-Object -First 1 -ExpandProperty AppID`,
      ],
      { encoding: "utf-8", timeout: 15000, windowsHide: true },
    );
    const appId = (r.stdout ?? "").trim();
    if (appId && r.status === 0) {
      const uri = `shell:AppsFolder\\${appId}`;
      appxCache.set(pattern, uri);
      return uri;
    }
  } catch {
    /* ignore */
  }
  appxCache.set(pattern, null);
  return null;
}

/** 解析工具启动命令（数组形式）；未找到抛错。 */
function resolveCommand(tool: LauncherToolId): string[] {
  if (tool === "cursor") {
    const exe = IS_WINDOWS ? which("cursor.cmd", "cursor") : which("cursor");
    if (exe) return [exe];
    throw new Error("cursor 命令未找到，请确认 Cursor 已安装且在 PATH 中");
  }

  if (tool === "codex-cli") {
    const exe = IS_WINDOWS ? which("codex.cmd", "codex") : which("codex");
    if (exe) return [exe];
    throw new Error("codex 命令未找到，请确认 Codex CLI 已安装 (npm i -g @openai/codex)");
  }

  if (tool === "codex-app") {
    if (IS_WINDOWS) {
      const exe = which("codex-app.cmd", "codex-app", "Codex.exe");
      if (exe) return [exe];
      const appx = findAppxApp("Codex");
      if (appx) return ["explorer.exe", appx];
    } else {
      const exe = which("codex-app", "Codex");
      if (exe) return [exe];
    }
    throw new Error("Codex App 未找到，请确认 Codex 桌面应用已安装");
  }

  if (tool === "windsurf") {
    if (IS_WINDOWS) {
      const exe = which("windsurf.cmd", "windsurf", "Windsurf.exe");
      if (exe) return [exe];
      const appx = findAppxApp("Windsurf");
      if (appx) return ["explorer.exe", appx];
    } else {
      const exe = which("windsurf", "Windsurf");
      if (exe) return [exe];
    }
    throw new Error("windsurf 命令未找到，请确认 Windsurf 已安装且在 PATH 中");
  }

  if (tool === "claude-code") {
    const exe = IS_WINDOWS ? which("claude.cmd", "claude") : which("claude");
    if (exe) return [exe];
    throw new Error(
      "claude 命令未找到，请确认 Claude Code 已安装 (npm i -g @anthropic-ai/claude-code)",
    );
  }

  if (tool === "claude-app") {
    if (IS_WINDOWS) {
      const exe = which("claude-app.cmd", "claude-app", "Claude.exe");
      if (exe) return [exe];
      const appx = findAppxApp("Claude");
      if (appx) return ["explorer.exe", appx];
    } else {
      const exe = which("claude-app", "Claude");
      if (exe) return [exe];
    }
    throw new Error("Claude App 未找到，请确认 Claude 桌面应用已安装");
  }

  throw new Error(`不支持的工具: ${tool}`);
}

/** 后台静默启动（GUI 应用）。 */
function launchBackground(cmd: string[]): void {
  const [bin, ...args] = cmd;
  const child = spawn(bin, args, {
    detached: true,
    stdio: "ignore",
    windowsHide: true,
  });
  child.unref();
}

/** 新终端窗口启动（CLI 交互式工具）。 */
function launchTerminal(cmd: string[]): void {
  if (IS_WINDOWS) {
    // start 一个新控制台运行命令（cmd /k 保留窗口）。
    const child = spawn("cmd.exe", ["/c", "start", "cmd", "/k", ...cmd], {
      detached: true,
      stdio: "ignore",
      windowsHide: false,
    });
    child.unref();
  } else if (IS_MAC) {
    const script = cmd.join(" ");
    spawn("osascript", [
      "-e",
      `tell app "Terminal" to do script "${script}"`,
    ]).unref();
  } else {
    for (const term of ["x-terminal-emulator", "gnome-terminal", "xterm"]) {
      if (which(term)) {
        spawn(term, ["-e", ...cmd], { detached: true, stdio: "ignore" }).unref();
        return;
      }
    }
    spawn(cmd[0], cmd.slice(1), { detached: true, stdio: "ignore" }).unref();
  }
}

/** 列出所有支持的工具及可用状态（对齐 GET /launcher/tools）。 */
export function listTools(): { tools: LauncherToolInfo[] } {
  const tools: LauncherToolInfo[] = SUPPORTED_TOOLS.map((id) => {
    let available = false;
    try {
      resolveCommand(id);
      available = true;
    } catch {
      available = false;
    }
    let mode: "app" | "terminal";
    let description: string;
    if (id === "codex-cli") {
      mode = "terminal";
      description = "在终端中启动 Codex CLI 交互式对话";
    } else if (id === "codex-app") {
      mode = "app";
      description = "启动 Codex 桌面应用";
    } else if (id === "windsurf") {
      mode = "app";
      description = "启动 Windsurf IDE";
    } else if (id === "claude-code") {
      mode = "terminal";
      description = "在终端中启动 Claude Code 交互式对话";
    } else if (id === "claude-app") {
      mode = "app";
      description = "启动 Claude 桌面应用";
    } else {
      mode = "app";
      description = "启动 Cursor IDE";
    }
    return { id, label: TOOL_LABELS[id], available, mode, description };
  });
  return { tools };
}

/** 启动工具（对齐 POST /launcher/launch）。 */
export function launchTool(
  req: LauncherLaunchRequest,
): LauncherLaunchResponse {
  const tool = req.tool;
  if (!SUPPORTED_TOOLS.includes(tool)) {
    throw new Error(
      `不支持的工具: ${tool}，可选: ${SUPPORTED_TOOLS.join(", ")}`,
    );
  }

  const cmd = resolveCommand(tool);
  if (req.project_path) {
    cmd.push(req.project_path);
  }

  const label = TOOL_LABELS[tool];
  const isTerminal = TERMINAL_TOOLS.has(tool);
  if (isTerminal) {
    launchTerminal(cmd);
  } else {
    launchBackground(cmd);
  }

  const suffix = req.project_path ? `，项目路径: ${req.project_path}` : "";
  return {
    status: "launched",
    tool,
    mode: isTerminal ? "terminal" : "app",
    message: `${label} 已成功启动${suffix}`,
  };
}
